7주차 - 레스토랑 로그인 페이지 및 리뷰 작성 페이지 구현하기

yxxnhx opened the pull request:
# 목표
레스토랑 로그인 페이지 및 리뷰 작성 페이지 구현하기

### 구현사항
- [x] 로그인 form 생성하기
- [x] 로그인 페이지 라우팅하기
- [x] 로그인하여 accessToken 받아오기
- [x] localStorage 활용하여 로그인 상태 유지하기
- [x] 로그아웃 기능 구현하기
- [x] 리뷰 추가 기능 구현하기
- [x] 리뷰 화면에 출력하기
- [x] 새로 입력한 리뷰가 업데이트 되기

### 진행사항
안녕하세요 트레이너님!
현재 강의 모두 듣고 로그인부터 차근차근히 하는 중입니다.
지금까지는 로그인을 하여 accessToken까지 받아온 상황입니다. 이제 localStorage 활용해서 로그인 상태 유지하는 것, 그리고 로그아웃 기능까지 구현하였습니다.
지금은 api에 리뷰를 추가하면 201이 뜨고 httpie를 이용해서 확인하면 추가되는 것까지 확인하였습니다.
이제 화면에 출력을 어떻게 할지 고민 중입니다.
레스토랑 상세 정보를 출력하였듯이 그렇게 하면 될 것 같기는 한데.. 한번 시도해보겠습니다!
API를 직접 이렇게 가져와서 로그인하고 추가하고 하는 것이 처음이라 많이 어렵네요..🥺
그럼 이번 한주도 잘 부탁드립니다!

yxxnhx committed:
App 컴포넌트에 LoginPage 연결

로그인 페이지를 App 컴포넌트에 Route를 활용하여 연결하였습니다

yxxnhx committed:
HomePage 컴포넌트에 Login 페이지 링크 연결

처음 접속 시, 링크로 들어가 로그인할 수 있도록 연결하였습니다

yxxnhx committed:
LoginPage 컴포넌트 생성 및 구현

페이지에서는 화면만 랜더링할 수 있도록 관심사를 분리하여 전체적인 상태는 container가 할 수 있도록 하였습니다

yxxnhx committed:
LoginFormContaienr 컴포넌트 생성 및 상태 관리 부여

컨테이너에서는 로그인에 관련한 handleSubmit, handleChange, accessToken을 가져오는 등 상태만 관리하고 직접적인 input은 Form에서 할 수 있도록 따로 분리하였습니다.

yxxnhx committed:
actions에 로그인할 수 있도록 requestLogin, 입력값을 받을 수 있도록 changeLoginField action을 추가

requestLogin에서는 getState로 현재 상태를 받아와서 postLogin으로 api를 호출한 후 accessToken을 받을 수 있도록 하였습니다.
changeLoginField에서는 입력하는 name과 value를 받아 상태를 관리할 수 있도록 하였습니다.

yxxnhx committed:
reducer 테스트에서 setAccessToken, changeLoginField action 검증

초기값에 loginFields, accessToken 빈값으로 추가 후,
setAccessToken으로 accessToken을 가져오는지,
changeLoginField에서는 각각 이메일만 입력하였을 때, 이메일만 바뀌고, 비밀번호를 입력하였을 때 비밀번호만 바뀌는지 여부를 검증하였습니다

yxxnhx committed:
reducer에서 setAccessToken, changeLoginField action 추가

setAccessToken에서 accessToken을 반환할 수 있도록, changeLoginField에서는 현재 loginFields 값을 풀어서 새롭게 입력한 값을 넣어줄 수 있도록
action을 추가하여 reducer 테스트를 통과하였습니다

yxxnhx committed:
api에 로그인할 수 있는 postLogin 생성

email과 password를 받아서 POST로 데이터를 넘겨 accessToken을 받아낼 수 있도록 하였습니다

yxxnhx committed:
fixtures 폴더에 테스트용 데이터 loginFields 생성

테스트용으로 로그인 데이터를 생성하였습니다

yxxnhx committed:
LoginForm 컴포넌트 생성 및 input 출력

input에 관한 form은 LoginForm 컴포넌트에서 전체적으로 관리할 수 있도록 분리하였습니다

yxxnhx committed:
LoginPage 테스트 생성 및 구현

로그인 페이지가 제대로 구현되는지 여부만 먼저 간단하게 테스트하였습니다

yxxnhx committed:
LoginFormContainer 테스트 생성 및 input이 랜더링 되는지, onChange 이벤트가 호출되는지 여부 테스트

상태를 관리하는 컨테이너에서 제대로 랜더링되는지, onChange 이벤트가 호출되는지 여부를 테스트하였습니다

yxxnhx committed:
LoginForm 테스트 파일 생성 및 form이 제대로 화면에 출력되는지 확인

form이 제대로 출력되는지, 입력한 값이 제대로 들어오는지 여부 확인하는 테스트를 구현하였습니다

yxxnhx committed:
RestaurantContaienr 컴포넌트 ReviewForm 컴포넌트 연결

accessToken이 있을 시에만 ReveiwForm이 보이도록 삼항연산자를 활용하여 설정하였습니다.

yxxnhx committed:
ReviewForm 컴포넌트 생성 및 평점과 리뷰 내용을 입력할 수 있도록 페이지 구현

ReviewForm에서는 화면 출력만 하고 실제 입력은 TextField에서 할 수 있도록 나누어 정리하였습니다

yxxnhx committed:
TextField 컴포넌트 생성 및 입력 화면 출력

상태는 RestaurantContainer, 출력할 정보는 ReviewForm에서 받아 화면을 출력만 할 수 있도록 정리하였습니다

yxxnhx committed:
fixtures에 테스트용 데이터 loginControls, reviewControls, reviewFields 생성

테스트에서 사용할 수 있도록 배열로 fixtures에 생성하였습니다

yxxnhx committed:
RestaurantContainer 테스트 로그인 여부에 따른 리뷰 form 랜더링과 이벤트 호출 여부 테스트

로그인 여부에 따라서 form이 랜더링되거나 되지 않거나를 나누어 테스트 진행하였습니다

yxxnhx committed:
RestaurantPage 테스트에 리뷰 작성 form 랜더링 여부 추가

리뷰 작성 기능이 추가됨으로 리뷰 작성 form이 함께 랜더링되는지 테스트 조건에 추가하였습니다

yxxnhx committed:
ReviewForm 테스트 파일 생성 및 리뷰 작성 form 랜더링 여부 테스트

리뷰 form과 onchange가 호출되는지, 리뷰 남기기 버튼이 랜더링 여부를 확인하는 테스트를 구현하였습니다

yxxnhx committed:
TextField 테스트 파일 생성 및 조건에 따른 form 랜더링 확인 테스트 구현

type의 기본값을 text로 주어 없을 경우 text로 랜더링되는지, 지정할 경우에는 지정한 값이 잘 랜더링 되는지 여부를 확인하는 테스트를 진행하였습니다

yxxnhx committed:
actions에 리뷰 관련 changeReviewField, sendReview, 로그아웃 기능 추가  생성

리뷰 작성 form에 입력 시, 입력 값이 들어가는 changeReviewField를 생성하였습니다.
입력한 리뷰를 api로 postReivew를 활용하여 보내는 sendReivew를 생성하였습니다.
로그아웃 기능을 위해 logout을 생성하였습니다

yxxnhx committed:
api에 작성한 리뷰를 보내는 postReivew 기능 구현

현재 restaurantId를 받아 경로를 지정하고, 입력한 score과 description을 넘길 수 있도록 accessToken을 활용하여 구현하였습니다

yxxnhx committed:
테스트용 api 생성

테스트용으로 postLogin, postReview를 빈값으로 반환하는 데이터를 생성하였습니다

yxxnhx committed:
reducer에 로그아웃 기능 logout, 리뷰 작성 changeReivewField 처리

logout에서는 빈 accessToken을 반환하여 리뷰 작성 form이 보이지 않도록 설정하였습니다
changeReviewField에서는 현재 state와 reviewField의 상태를 반환하여 각각 입력 시 기존 값이 변하지 않도록 설정하였습니다

yxxnhx committed:
reducer 테스트에 logout, changeReivewField 조건에 따른 테스트 구현

logout 시, 빈 accessToken을 반환하는지 여부를 테스트하였습니다.
changeReivewField에서는 평점 혹은 리뷰 둘 중 하나를 입력 시 각각 입력한 값만 변경되도록 설정하였습니다

yxxnhx committed:
로그아웃 기능 추가를 위해 LoginFormContainer 컴포넌트에 accessToken 여부에 따른 조건 추가

accessToken이 있을 경우에만 LogoutForm이 보이고, 아닐 시에는 LoginForm만 출력되도록 조건을 추가하였습니다

yxxnhx committed:
LoginFormContainer 테스트에 로그인 여부에 따른 조건 추가

로그인을 하지 않았거나 혹은 로그아웃을 했을 경우, 로그인을 하였을 경우에 따른 테스트 조건을 추가하였습니다

yxxnhx committed:
LogoutForm 컴포넌트 생성 및 로그아웃 버튼 구현

로그아웃 버튼을 화면에만 출력할 수 있도록 관심사를 분리하여 구현하였습니다

yxxnhx committed:
LogoutForm 테스트 파일 생성 및 로그아웃 페이지의 랜더링 여부 테스트 구현

버튼을 랜더링하고 버튼 클릭 시, handleClick이 호출되는지 여부를 확인하였습니다

yxxnhx committed:
App컴포넌트에 localStoarge를 활용하여 로그인 상태 저장 추가

새로고침 시 로그인이 풀리던 이슈를 localStorage를 활용하여 상태를 저장할 수 있도록 setAccessToken을 활용하여 기능을 추가하였습니다

yxxnhx committed:
App 테스트에 로그인 여부에 따른 테스트 조건 추가

로그인을 하였을 경우 디스패치를 호출하고, 로그아웃 시에는 호출하지 못하는 조건을 추가하였습니다

yxxnhx committed:
services에 localStorage를 활용하기 위해 storage 생성

getItem, setItem을 하고 테스트 시 조금 더 용이하게 할 수 있도록 따로 빼서 정리하였습니다

yxxnhx committed:
테스트용 localStorage 데이터 생성

테스트 시 localStorage를 쉽게 사용하기 위하여 테스트용 데이터를 생성하였습니다

yxxnhx committed:
LoginForm에 필요없는 코드 삭제

현재 handleChange 내에서는 preventDefault가 필요없어 삭제하였습니다

yxxnhx committed:
LoginForm 테스트에 테스트용 데이터를 활용하여 코드 정리

테스트용 데이터 loginControls를 활용하여 중복된 데이터 정리하였습니다

daadaadaah reviewed (commented):
이야기 안했는데, 스스로 given 도 쓰시고 정말 잘하셨어요!👍 
다른거 잘 구현하셨는데, 로그아웃 기능 구현할 때 localStorage도 어떤 처리 필요하지 않을까요?ㅎㅎ
다음 PR 기다리고 있을게요!🚀

yxxnhx committed:
로그아웃 시 localStorage의 accessToken 삭제 기능 추가

removeItem을 이용하여 localStorage에 저장된 accessToken을 삭제할 수 있도록 추가하였습니다

yxxnhx committed:
테스트용 데이터 deleteItem 추가

테스트할 시 localStorage를 용이하게 사용하기 위하여 추가하였습니다

yxxnhx committed:
LoginFormContaienr에 로그아웃 시 localStorage에 저장된 accessToken 삭제 기능 추가

removeItem을 활용하여 만든 deleteItem을 handleClickLogout에 추가하여 로그아웃 후 새로고침을 하였을 때 여전히 남아 다시 로그인 되던 에러를 해결했습니다

yxxnhx committed:
Reviews 컴포넌트 생성 및 리뷰 화면 출력 구현

reviews를 받아 화면에 최신 등록부터 각각 이름, 점수, 평가가 각각 출력될 수 있도록 구현하였습니다.

yxxnhx committed:
Reviews 테스트 파일 생성 및 리뷰 랜더링 여부 확인

리뷰 데이터의 유무에 따른 화면 출력 여부 테스트를 확인하였습니다

yxxnhx committed:
화면 출력 순서 정리를 위해 RestaurantContainer 컴포넌트에서 ReviewForm 삭제

메뉴 아이템 아래에 바로 ReviewForm이 나오는 것으로 구현하기 위하여 RestaurantDetail에서 관리할 수 있도록 관련 데이터들은 RestaurantDetail 컴포넌트에 넘길 수 있도록 코드를 정리하였습니다

yxxnhx committed:
RestaurantDetail 컴포넌트에서 ReviewForm의 랜더링 여부 관리할 수 있도록 수정 및 Reviews 컴포넌트 연결

accessToken의 여부에 따라 ReveiwForm의 랜더링 여부를 결정할 수 있도록 삼항연산자를 이용하여 구현하였습니다.
Reveiws 컴포넌트를 연결하여 메뉴 아이템 하단에 Review가 보일 수 있도록 하였습니다

yxxnhx committed:
테스트용 데이터 reviews 생성

리뷰에 관한 내용을 테스트할 경우 사용할 수 있게 fixtures에 테스트용 리뷰 데이터를 생성하였습니다

yxxnhx committed:
리뷰를 보낸 후 다시 한번 레스토랑이 랜더링 되도록 loadRestaurant 추가

리뷰 작성 후 새로고침 해야만 입력한 데이터가 나오는 이슈를 loadRestaurant를 추가하여 해결하였습니다

yxxnhx commented:
# 진행사항
로그아웃 시 localStorage에 저장되었던 accessToken을 삭제할 수 있도록 기능을 추가하였습니다
그리고 리뷰들을 화면에 출력하였고 리뷰 작성 시 자동으로 loadRestaurant를 하여 입력한 리뷰가 상단에 보일 수 있도록 reverse를 이용하여 역순으로 나올 수 있게 하였습니다!

### 질문사항
![image](https://user-images.githubusercontent.com/50559373/203782835-7a12d8d8-b9d9-4897-98ca-7684fee3d59a.png)
```jsx
export default function Reviews({ reviews }) {
  if (!reviews || reviews.length === 0) {
    return <p>리뷰가 없어요!</p>;
  }
  return (
    <ul>
      {
        reviews.reverse().map((review) => (
          <li key={review.id}>
            <p>{review.name}</p>
            <p>{review.score}</p>
            <p>{review.description}</p>
          </li>
        ))
      }
    </ul>
  );
}
```
이미 Reviews에는 key가 있는데 왜 이러한 에러가 뜰까요?
로컬 크롬 콘솔창에는 위와 같은 에러가 뜨지 않습니다ㅠㅠ

![Nov-24-2022 21-21-16](https://user-images.githubusercontent.com/50559373/203783522-79dea854-ab41-4afe-b643-49a62ff112a1.gif)
그리고 리뷰를 입력할 때 자꾸 아래 리뷰들이 바뀝니다. 이게 어떠한 문제로 일어나는지를 모르겠어서 해결하지 못하고 있습니다ㅠㅠ
어디를 살펴봐야 할까요?

daadaadaah commented on src/LoginFormContainer.jsx:
```diff
@@ -26,6 +27,7 @@ export default function LoginFormContainer() {
 
   function handleClickLogout() {
     dispatch(logout());
+    deleteItem('accessToken');
```
`LoginFormContainer`가 services를 모르게 하기 위해서 accessToken 삭제 로직을 어디서 관리하는게 좋을까요?

yxxnhx replied:
actions에 추가하도록 하겠습니다!

daadaadaah commented on fixtures/reviews.js:
```diff
@@ -0,0 +1,14 @@
+const reviews = [
+  {
+    name: '테스터',
+    score: '5',
+    description: '맛있어요👍',
+  },
+  {
+    name: '테스터',
+    score: '1',
+    description: '우욱',
+  },
+];
```
테스트 코드에 문제가 생겼을 때에는 테스트 코드에서 그 원인을 찾는게 좋아요!
지금 사용하고 계신 fixture로 사용하고 있는 reviews 데이터와 실제 구현 코드에서 사용하고 있는 review 데이터에 어떤 차이가 있을까요?

yxxnhx replied:
앗 데이터에 id가 없어서 일어난 것이엇군요

daadaadaah reviewed (changes requested):
질문에 대한 답변은 리뷰에 달아놨어요! 수정하실 때 참고하세요!ㅎㅎ

yxxnhx committed:
Reviews 테스트의 key 부재로 인한 에러 해결

테스트용 데이터 reviews에 id가 없어서 일어난 에러로 id를 추가하여 에러 해결하였습니다

yxxnhx committed:
리뷰를 입력할 때마다 계속해서 리뷰들이 재로딩되는 문제 해결

api에서 받은 리뷰를 그대로 reverse 사용해서 일어난 에러로 spread 연산자를 이용해서 풀어서 사용할 수 있도록 수정하여 에러 해결하였습니다

yxxnhx committed:
reducer 초기값 순서 수정

가장 중요한 accessToken을 상단으로 올려 순서를 수정하였습니다

yxxnhx committed:
actions 테스트에 postLogin, sendReivew 추가

postLogin을 호출할 시, setAccessToken을 디스패치 하여 값이 들어오는지 여부를 확인하였습니다
sendReivew를 호출할 시, setRestaurant를 디스패치하여 restaurantId에 맞는 레스토랑이 나오는지 여부를 확인하였습니다

yxxnhx committed:
api의 postReview 반환 형식 설정 수정

바로 response를 반환하지 말고 data로 받아서 반환할 수 있도록 수정하였습니다

yxxnhx committed:
api 테스트에 postLogin, postReviw 테스트 추가

postLogin을 할 경우 accessToken을 반환하는지 여부를 테스트하였습니다.
postReivew를 할 경우 리뷰를 반환하는지 확인하는 테스트를 하였습니다.

yxxnhx committed:
storage 테스트 파일 생성 및 구현

localStorage의 데이터를 가져오고 삭제하고 넣을 수 있는지 여부를 확인하였습니다

yxxnhx committed:
eslint로 코드 수정

yxxnhx committed:
관심사 분리에 따른 LoginFormContainer 컴포넌트에서 services 삭제

localStorage에서 accessToken을 삭제하는 것은 actions의 logout에서 처리할 수 있도록 해당 코드 삭제하였습니다

yxxnhx committed:
actions의 logout에서 localStorage의 accessToken 삭제 기능 추가

관심사 분리를 위해 LoginFormContaier 컴포넌트에서 실행하던 deleteItem을 actions로 옮겨서 수정하였습니다

yxxnhx committed:
e2e 테스트 통과를 위해 Log in 문구를 Log In으로 수정

yxxnhx committed:
e2e 테스트 통과를 위해 Log in 문구를 Log In으로 수정

yxxnhx commented:
# 진행사항
관심사 분리를 위해 services를 LoginFormContainer가 직접 알고 있지 않아도 되게 actions에 추가하였습니다.
그리고 테스트 파일에 id가 없어서 테스트에서만 에러가 뜨는 것 또한 추가하여 해결하였습니다.
리뷰 값을 입력할 때마다 계속해서 리뷰가 재랜더링 되어 삐그덕거리는 것처럼 보이는 것은 스프래드 연산자를 이용하여 받아온 리뷰를 바로 사용하지 않고 풀어서 사용할 수 있도록 하여 테스트 커버리지도 통과하였습니다.

yxxnhx committed:
sendReview 시 리뷰만 로드, reviewFields 초기화 기능 추가 수정

loadReview를 추가하여 setReivew를 디스패치하여서 리뷰만 로드될 수 있도록 로직을 수정하였습니다
리뷰 남기기를 누르면 기존에 입력한 input창도 지워질 수 있도록 수정하였습니다

yxxnhx committed:
reducer에 setReviews, clearReviewFields 추가

리뷰를 가져오고, 리뷰 input 창을 초기화시켜주는 action을 추가하였습니다

yxxnhx committed:
setReviews, clearReviewFields 테스트 추가

리뷰를 가져와서 변화시켜주는지 여부를 테스트하였습니다
review field를 초기화시켜서 비워지는지 여부를 확인하였습니다

yxxnhx committed:
RestaurantContainer에서 RestaurantDetail에 reviewFields 넘기는 형식을 수정

기존에는 reviewFields는 reviewFields로 넘겼으나 간소화하기 위해 fields로 넘기는 식으로 수정하였습니다

yxxnhx committed:
RestaurantDetial에서 fields로 수정

fields를 props 내릴 수 있도록 하였씁니다

yxxnhx committed:
ReviewForm에서 fields를 받아 값이 들어오는 것을 볼 수 있도록 수정

기존에는 fields를 받은 것을 활용하지 않았으나 value에 넣어 각각 값을 관리하여 초기화까지 할 수 있도록 수정하였습니다

yxxnhx committed:
TextField 컴포넌트에서 value 받을 수 있도록 수정

ReviewForm에서 value를 props 받아서 값을 받을 수 있도록 수정하였습니다

yxxnhx committed:
TextField 테스트에 입력 값 랜더링 여부 테스트 추가

컴포넌트에서 입력한 값을 랜더링하는지 여부 테스트를 진행하였습니다

yxxnhx committed:
ReviewForm 테스트에 review field 랜더링 여부 추가

리뷰 field에 입력한 값이 들어오는지 테스트 여부를 추가하였습니다

yxxnhx committed:
actions 테스트에 loadReview와 sendReview 조건 추가

loadReview를 추가하여 setReview를 호출하는지 여부를 확인하였습니다
sendReview에서 clearReviewFields가 호출되는지 여부를 추가하였습니다

yxxnhx commented:
# 진행사항
해설 강의를 듣고 입력시 input 창이 초기화 되는 기능과 레스토랑 전체 대신 리뷰만 로드될 수 있도록 수정하였습니다
그런데 그 이전까지는 e2e 테스트가 모두 통과했었는데 지금 이상하게 테스트가 깨집니다
카테고리와 레스토랑 모두 선택하면 레스토랑 목록이 나오지 않는다고 하는데 왜일까요?

yxxnhx committed:
TextField 테스트 기본값 value 추가

value를 props 받아 내리기 때문에 render 함수에 value를 추가하였습니다

yxxnhx committed:
페이지별 파일 정리

LoginPage, RestaurnatPage, RestaurantsPage 별 폴더를 만들어서 파일을 정리하였습니다

yxxnhx commented:
## 진행사항
해설 강의를 듣고 입력시 input 창이 초기화 되는 기능과 레스토랑 전체 대신 리뷰만 로드될 수 있도록 수정하였습니다
파일들이 많아지면서 페이지별로 폴더를 정리하였는데 이 방법이 맞을까요? 아니면 조금 더 새분화해서 RestaurantPage 같은 경우엔 안에서 카테고리, 레스토랑, 지역을 나누어서 해주는 것이 좋을까요?

daadaadaah reviewed (approved):
> 파일들이 많아지면서 페이지별로 폴더를 정리하였는데 이 방법이 맞을까요? 아니면 조금 더 새분화해서 RestaurantPage 같은 경우엔 안에서 카테고리, 레스토랑, 지역을 나누어서 해주는 것이 좋을까요?

- 저도 폴더 및 파일 구조는 매 프로젝트 고민이 많이 되는 부분 같아요! 그래서, 저는 일단 한가지 방법으로 해보고, 그렇게 사용해보다가 불편하면 바꾸는 편입니다!ㅎㅎ페이지별로 나눠서 관리하는 것도 좋은 것 같아요! 이번주는 페이지별로 해보고, 다음주 과제 때에는 다른 방법으로 해보면서 각각의 장단점을 몸소 느껴보는게 도움이될 것 같습니다!ㅎㅎ이번주 한주도 수고 많으셨어요!👏
